### PR TITLE
[Domain Management] Domain Details web view

### LINKED ISSUE #21431: 🍎  Domain Details Screen as WebView
A webview that displays the domain details.

## Screenshot Reference

Image

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes #21431 and #21975

## Related PR
This PR depends on:
- https://github.com/wordpress-mobile/WordPressKit-iOS/pull/642

## Description

This PR adds the ability to view the domain details page in a web view.

| iPhone | iPad | 
| ------ | ----- |
|  |  |

## Test Instructions

1. Install Jetpack and login.
2. Enable `Domain Management` feature flag. 
3. Head to Me > All Domains.
4. Tap a domain.
5. **Expect** a navigation push to Domain Management web view.
6. Tap any button in that web page, for example "Payment Details".
7. **Expect** the new page to open in device browser.
8. Switch back to Jetpack.
9. **Expect** the Domain Management web page is still visible. 

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
None.

3. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Me/All Domains/Views/AllDomainsListViewController.swift
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Me/All Domains/Views/AllDomainsListViewController.swift:86]: This change is unrelated to this PR and it should've been added in a prior PR.

File: WordPress/Classes/ViewRelated/Me/All Domains/Views/AllDomainsListViewController+Strings.swift
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Me/All Domains/Views/AllDomainsListViewController+Strings.swift:14]: This change is unrelated to this PR and it should've been added in a prior PR.

File: WordPress/Classes/ViewRelated/Me/All Domains/View Models/AllDomainsListItemViewModel.swift
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Me/All Domains/View Models/AllDomainsListItemViewModel.swift]: This change is unrelated to this PR and it should've been added in a prior PR.

File: WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift
Comment (salimbraksa) [inline WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift:9]: The test cases were designed in a way to cover all lines of code in `AllDomainsListItemViewModel`. To verify that, enable [Code Coverage](https://www.swiftbysundell.com/tips/gathering-test-coverage-in-xcode/) then run the tests.

Comment (salimbraksa) [reply WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift:9]: @Mokagio, `AllDomainsListItemViewModel.swift` was originally intended for the Jetpack target. However, I had to include it in the WordPress target to run unit tests. Is there a way to execute tests specifically on the Jetpack target?

Comment (mokagio) [reply WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift:9]: @salimbraksa fun fact I _just_ posted in our internal devs channel a link to a PR to explore ways to go in that direction.

The PR is https://github.com/wordpress-mobile/WordPress-iOS/pull/21962 but, unfortunately, it doesn't make what you ask possible. So I guess we'll have to keep the file in both target for a little while.

Thank you for adding tests for this!

File: WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift
Comment (mokagio) [inline WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift]: Nitpick. In XCTest, functions that start with "test" are picked up by the test runner. I don't think this will because it has arguments, too, but IMHO it would be clearer if it had a different name. For example given that it wraps an assertion:

```suggestion
    private func assert(domain: Domain, mapsTo expectedViewModel: ViewModel) {
        XCTAssertEqual(ViewModel(domain: domain), expectedViewModel)
    }
```

File: WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift]: How about we allow URLs that start with `https://wordpress.com/domains/manage/all/{DOMAIN}`?
This way, some simple actions like managing DNS records can be handled within the webview.

File: WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift:45]: [Suggestion] Instead of observing the url and navigating back, how about using `WKWebView.navigationDelegate`?
See [BlazeWebViewController.swift](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Blaze/Webview/BlazeWebViewController.swift) for reference

File: WordPress/Classes/ViewRelated/Me/All Domains/Views/AllDomainsListViewController.swift
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Me/All Domains/Views/AllDomainsListViewController.swift]: [Suggestion] This can make testing with a sandbox easier
```suggestion
        destination.configureSandboxStore {
            navigationController.pushViewController(destination, animated: true)
        }
```

File: WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift
Comment (hassaanelgarem) [inline WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift]: [Nit] Consider moving this to the beginning of the file

**Review by hassaanelgarem — COMMENTED** [review]
Works as expected with one small exception:

The nav bar title should show the domain's title instead of reacting to the webview title. This can be achieved through `WebKitViewController.configuration.customTitle`. However, there's a bug in `WebKitViewController` which I fixed in #21979 in https://github.com/wordpress-mobile/WordPress-iOS/pull/21979/commits/536cdcd3ec1bcf56b0bcdb3303746a0b1bfc5d00. You can cherry-pick this or wait for this PR to be merged 👍

File: WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift:45]: @hassaanelgarem That was the first thing I've tried, but it didn't catch all URL changes. For example, that method wasn't called when I tapped `Payment Details` button.

I also tried overriding this `WKNavigationDelegate` method:

```swift
func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void)
```

Navigation info can be in these properties:
- `navigationAction.request.url `
- `navigationAction.sourceFrame.request.url`
- `navigationAction.targetFrame.request.url`

I tried all those properties but there was still cases where this delegate method doesn't catch all URL changes.

File: WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift
Comment (salimbraksa) [reply WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift]: Resolved in ec8683b42259a0f2aa2dc95d5400356c262ab639

File: WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift
Comment (salimbraksa) [reply WordPress/WordPressTest/Domains/AllDomainsListItemViewModelTests.swift]: Resolved in bce644663ff0e8bd1fa063fc32c48ead9854794a

File: WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Me/Domain Details/DomainDetailsWebViewController.swift]: Hey @hassaanelgarem, I've pushed multiple commits implementing this suggestion.

**Comment by salimbraksa** [comment]
> The nav bar title should show the domain's title instead of reacting to the webview title.

@hassaanelgarem Addressed in 83f15816168508d08febef91057892b1c23890e3. But as you said, it's not going to work because of a bug in `customTitle`. Since this issue is not blocking, what do you say we merge this PR, then this issue will be resolved once your PR is merged into `trunk`?

**Review by hassaanelgarem — COMMENTED** [review]
I noticed the following issues after enabling more navigation inside the webview:

## Multiple back buttons
When navigating to a sub page like DNS records, the webview displays a back button. I think it's weird having two back buttons, one in the nav bar and one in the web view and both doing different things.

What do you think if we display the webview as a modal instead of pushing it? We can have a done button on the right side of the nav bar. This way there will be a clear distinctions between the two user actions.



## Connecting domain to existing site
When connecting a domain to an existing site, the flow works great up until completion. Once completed, there's weird behavior. The user is navigated to the device's brower and back in the app, the webview is empty and not interactable. Trying to re-open the domain details doesn't work. The user has to refresh the domains list to get the domain details web view to work again.

Can we detect the completion state and dismiss the webview and reload the list?


https://github.com/wordpress-mobile/WordPress-iOS/assets/25306722/faf598eb-5497-4a24-8ef8-9b37d786189a

**Comment by hassaanelgarem** [comment]
> what do you say we merge this PR, then this issue will be resolved once your PR is merged into trunk?

@salimbraksa Yes, that makes sense. Can we create an issue to track this, though? Once both PRs are merged, we can just confirm the issue is fixed and close the issue. What do you think?

**Comment by salimbraksa** [comment]
> Yes, that makes sense. Can we create an issue to track this, though? Once both PRs are merged, we can just confirm the issue is fixed and close the issue. What do you think?

@hassaanelgarem I think that won't be necessary, your [PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/21979) was just merged, so I'll sync my branch with `trunk`.

**Comment by salimbraksa** [comment]
> Multiple back buttons

@hassaanelgarem I'd suggest to align with Android on this. See Slack discussion. p1699522050271729/1699462442.594229-slack-C05NS0YV7HS

> Connecting domain to existing site

Resolved in da7b4300381c8497b72179098475104ac95e2bb6. A similar issue happens when a domain is deleted, but I couldn't figure out a way to catch this event.

**Comment by salimbraksa** [comment]
Hey @hassaanelgarem I've reverted the changes to preventing all navigation within the web view.

**Review by hassaanelgarem — APPROVED** [review]
🚀

**Pull request merged**
